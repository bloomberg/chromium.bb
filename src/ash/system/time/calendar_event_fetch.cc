// Copyright 2022 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "ash/system/time/calendar_event_fetch.h"

#include "ash/calendar/calendar_client.h"
#include "ash/calendar/calendar_controller.h"
#include "ash/shell.h"
#include "ash/system/time/calendar_utils.h"
#include "base/bind.h"
#include "base/check.h"
#include "base/metrics/histogram_functions.h"
#include "base/time/tick_clock.h"
#include "base/time/time.h"

namespace ash {

CalendarEventFetch::CalendarEventFetch(
    const base::Time& start_of_month,
    FetchCompleteCallback complete_callback,
    FetchInternalErrorCallback internal_error_callback,
    const base::TickClock* tick_clock)
    : start_of_month_(start_of_month),
      complete_callback_(std::move(complete_callback)),
      internal_error_callback_(std::move(internal_error_callback)),
      fetch_start_time_(base::Time::Now()),
      timeout_(tick_clock) {
  CalendarClient* client = Shell::Get()->calendar_controller()->GetClient();
  DCHECK(client);

  const base::Time start_of_next_month =
      calendar_utils::GetStartOfNextMonthUTC(start_of_month);
  client->GetEventList(base::BindOnce(&CalendarEventFetch::OnResultReceived,
                                      weak_factory_.GetWeakPtr()),
                       start_of_month, start_of_next_month);

  timeout_.Start(
      FROM_HERE, calendar_utils::kEventFetchTimeout,
      base::BindOnce(&CalendarEventFetch::OnTimeout, base::Unretained(this)));
}

CalendarEventFetch::~CalendarEventFetch() = default;

void CalendarEventFetch::OnResultReceived(
    google_apis::ApiErrorCode error,
    std::unique_ptr<google_apis::calendar::EventList> events) {
  // Cancel timeout timer.
  timeout_.Stop();

  base::UmaHistogramTimes("Ash.Calendar.FetchEvents.FetchDuration",
                          base::Time::Now() - fetch_start_time_);
  base::UmaHistogramBoolean("Ash.Calendar.FetchEvents.Timeout", false);

  // IMPORTANT: 'this' is NOT safe to use after `complete_callback_` has been
  // executed, as the last thing it does is destroy its
  // std::unique_ptr<CalendarEventFetch> to this object.
  std::move(complete_callback_).Run(start_of_month_, error, events.get());
}

void CalendarEventFetch::OnTimeout() {
  base::UmaHistogramBoolean("Ash.Calendar.FetchEvents.Timeout", true);

  // IMPORTANT: 'this' is NOT safe to use after `internal_error_callback_` has
  // been executed, as the last thing it does is destroy its
  // std::unique_ptr<CalendarEventFetch> to this object.
  std::move(internal_error_callback_)
      .Run(start_of_month_, CalendarEventFetchInternalErrorCode::kTimeout);
}

}  // namespace ash
