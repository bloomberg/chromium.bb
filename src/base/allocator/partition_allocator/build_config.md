# Build Config

PartitionAlloc's behavior and operation can be influenced by many
different settings. Broadly, these are controlled at the top-level by
[GN args][gn-declare-args], which propagate via
[buildflags][buildflag-header] and `#defined` clauses.

*** promo
Most of what you'll want to know exists between

* [`//base/allocator/partition_allocator/BUILD.gn`][pa-build-gn],
* [`allocator.gni`][allocator-gni],
* [`//base/allocator/BUILD.gn`][base-allocator-build-gn], and
* [`//base/BUILD.gn`][base-build-gn].
***

*** aside
While Chromium promotes the `#if BUILDFLAG(FOO)` construct, some of
PartitionAlloc's behavior is governed by compound conditions `#defined`
in [`partition_alloc_config.h`][partition-alloc-config].
***

## Select GN Args

### `use_partition_alloc`

Defines whether PartitionAlloc is at all available.

Setting this `false` will entirely remove PartitionAlloc from the
Chromium build. _You probably do not want this._

*** note
Back when PartitionAlloc was the dedicated allocator in Blink, disabling
it was logically identical to wholly disabling it in Chromium. This GN
arg organically grew in scope with the advent of
PartitionAlloc-Everywhere and must be `true` as a prerequisite for
enabling PA-E.
***

### `use_allocator`

Does nothing special when value is `"none"`. Enables
[PartitionAlloc-Everywhere (PA-E)][pae-public-doc] when value is
`"partition"`.

*** note
* While "everywhere" (in "PartitionAlloc-Everywhere") tautologically
  includes Blink where PartitionAlloc originated, setting
  `use_allocator = "none"` does not disable PA usage in Blink.
* `use_allocator = "partition"` internally sets
  `use_partition_alloc_as_malloc = true`, which must not be confused
  with `use_partition_alloc` (see above).
***

### `use_backup_ref_ptr`

Specifies `BackupRefPtr` as the implementation for `base::raw_ptr<T>`
when `true`. See the [MiraclePtr documentation][miracleptr-doc].

*** aside
BRP requires support from PartitionAlloc, so `use_backup_ref_ptr` also
compiles the relevant code into PA. However, this arg does _not_ govern
whether or not BRP is actually enabled at runtime - that functionality
is controlled by a Finch flag.
***

## Note: Component Builds

When working on PartitionAlloc, know that `is_debug` defaults to
implying `is_component_build`, which interferes with the allocator
shim. A typical set of GN args should include

```none
is_debug = true
is_component_build = false
```

Conversely, build configurations that have `is_component_build = true`
without explicitly specifying PA-specific args will not build with PA-E
enabled.

[gn-declare-args]: https://gn.googlesource.com/gn/+/refs/heads/main/docs/reference.md#func_declare_args
[buildflag-header]: https://source.chromium.org/chromium/chromium/src/+/main:build/buildflag_header.gni
[pa-build-gn]: https://source.chromium.org/chromium/chromium/src/+/main:base/allocator/partition_allocator/BUILD.gn
[allocator-gni]: https://source.chromium.org/chromium/chromium/src/+/main:base/allocator/allocator.gni
[base-allocator-build-gn]: https://source.chromium.org/chromium/chromium/src/+/main:base/allocator/BUILD.gn
[base-build-gn]: https://source.chromium.org/chromium/chromium/src/+/main:base/BUILD.gn
[partition-alloc-config]: https://source.chromium.org/chromium/chromium/src/+/main:base/allocator/partition_allocator/partition_alloc_config.h
[pae-public-doc]: https://docs.google.com/document/d/1R1H9z5IVUAnXJgDjnts3nTJVcRbufWWT9ByXLgecSUM/preview
[miracleptr-doc]: https://docs.google.com/document/d/1pnnOAIz_DMWDI4oIOFoMAqLnf_MZ2GsrJNb_dbQ3ZBg/preview
