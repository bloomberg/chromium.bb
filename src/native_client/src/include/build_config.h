/*
 * Copyright 2014 The Native Client Authors. All rights reserved.
 * Use of this source code is governed by a BSD-style license that can be
 * found in the LICENSE file.
 */

/*
 * All macros defined in this file should be checked by the presubmit check in
 * tools/code_hygiene.py to ensure that this header is included in source files
 * when needed.
 */

#ifndef NATIVE_CLIENT_SRC_INCLUDE_BUILD_CONFIG_H_
#define NATIVE_CLIENT_SRC_INCLUDE_BUILD_CONFIG_H_ 1

#include "native_client/src/include/nacl_base.h"

#if !defined(NACL_WINDOWS) && !defined(NACL_LINUX) && !defined(NACL_OSX) && \
    !defined(NACL_ANDROID)

#if defined(_WIN32)
# define NACL_WINDOWS 1
#else
# define NACL_WINDOWS 0
#endif

#if defined(__linux__)
# define NACL_LINUX 1
#else
# define NACL_LINUX 0
#endif

#if defined(__APPLE__)
# define NACL_OSX 1
#else
# define NACL_OSX 0
#endif

#if defined(ANDROID)
# define NACL_ANDROID 1
#else
# define NACL_ANDROID 0
#endif

#endif

/* TODO(teravest): Remove this guard when builds stop defining these. */
#if defined(NACL_BUILD_ARCH) && NACL_ARCH(NACL_BUILD_ARCH) != NACL_pnacl && \
    !defined(NACL_BUILD_SUBARCH)
# error Please define both NACL_BUILD_ARCH and NACL_BUILD_SUBARCH.
#elif !defined(NACL_BUILD_ARCH) && defined(NACL_BUILD_SUBARCH)
# error Please define both NACL_BUILD_ARCH and NACL_BUILD_SUBARCH.
#endif


#if !defined(NACL_BUILD_ARCH)

#if defined(_M_X64) || defined(__x86_64__)
# define NACL_BUILD_ARCH x86
# define NACL_BUILD_SUBARCH 64
#endif

#if defined(_M_IX86) || defined(__i386__)
# define NACL_BUILD_ARCH x86
# define NACL_BUILD_SUBARCH 32
#endif

#if defined(__ARMEL__)
# define NACL_BUILD_ARCH arm
# define NACL_BUILD_SUBARCH 32
#endif

#if defined(__MIPSEL__)
# define NACL_BUILD_ARCH mips
# define NACL_BUILD_SUBARCH 32
#endif

#endif  /* !defined(NACL_BUILD_ARCH) */

/*
 * TODO(teravest): Require NACL_BUILD_ARCH and NACL_BUILD_SUBARCH to be defined
 * once they're defined for the pnacl translator build.
 */

#endif  /* NATIVE_CLIENT_SRC_INCLUDE_BUILD_CONFIG_H_ */
