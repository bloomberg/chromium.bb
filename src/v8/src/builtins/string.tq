// Copyright 2019 the V8 project authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include 'src/builtins/builtins-string-gen.h'

namespace string {
  // ES6 #sec-string.prototype.tostring
  transitioning javascript builtin
  StringPrototypeToString(implicit context: Context)(receiver: Object): Object {
    return ToThisValue(receiver, kString, 'String.prototype.toString');
  }

  // ES6 #sec-string.prototype.valueof
  transitioning javascript builtin
  StringPrototypeValueOf(implicit context: Context)(receiver: Object): Object {
    return ToThisValue(receiver, kString, 'String.prototype.valueOf');
  }

  extern macro StringBuiltinsAssembler::LoadSurrogatePairAt(
      String, intptr, intptr, constexpr UnicodeEncoding): int32;
  extern macro StringFromSingleCodePoint(int32, constexpr UnicodeEncoding):
      String;

  // This function assumes StringPrimitiveWithNoCustomIteration is true.
  transitioning builtin StringToList(implicit context: Context)(string: String):
      JSArray {
    const kind = PACKED_ELEMENTS;
    const stringLength: intptr = string.length_intptr;

    const map: Map = LoadJSArrayElementsMap(kind, LoadNativeContext(context));
    const array: JSArray = AllocateJSArray(
        kind, map, stringLength, SmiTag(stringLength),
        kAllowLargeObjectAllocation);
    const elements = UnsafeCast<FixedArray>(array.elements);
    const encoding = UTF16;
    let arrayLength: Smi = 0;
    let i: intptr = 0;
    while (i < stringLength) {
      const ch: int32 = LoadSurrogatePairAt(string, stringLength, i, encoding);
      const value: String = StringFromSingleCodePoint(ch, encoding);
      elements[arrayLength] = value;
      // Increment and continue the loop.
      i = i + value.length_intptr;
      arrayLength++;
    }
    assert(arrayLength >= 0);
    assert(SmiTag(stringLength) >= arrayLength);
    array.length = arrayLength;

    return array;
  }

  transitioning macro GenerateStringAt(implicit context: Context)(
      receiver: Object, position: Object, methodName: constexpr string):
      never labels IfInBounds(String, intptr, intptr),
      IfOutOfBounds {
    // Check that {receiver} is coercible to Object and convert it to a String.
    const string: String = ToThisString(receiver, methodName);
    // Convert the {position} to a Smi and check that it's in bounds of
    // the {string}.
    const indexNumber: Number =
        ToInteger_Inline(context, position, kTruncateMinusZero);
    if (TaggedIsNotSmi(indexNumber)) goto IfOutOfBounds;
    const index: intptr = SmiUntag(UnsafeCast<Smi>(indexNumber));
    const length: intptr = string.length_intptr;
    if (Convert<uintptr>(index) >= Convert<uintptr>(length)) goto IfOutOfBounds;
    goto IfInBounds(string, index, length);
  }

  // ES6 #sec-string.prototype.charat
  transitioning javascript builtin StringPrototypeCharAt(
      implicit context: Context)(receiver: Object, position: Object): Object {
    try {
      GenerateStringAt(receiver, position, 'String.prototype.charAt')
          otherwise IfInBounds, IfOutOfBounds;
    }
    label IfInBounds(string: String, index: intptr, length: intptr) {
      const code: int32 = StringCharCodeAt(string, index);
      return StringFromSingleCharCode(code);
    }
    label IfOutOfBounds {
      return kEmptyString;
    }
  }

  // ES6 #sec-string.prototype.charcodeat
  transitioning javascript builtin StringPrototypeCharCodeAt(
      implicit context: Context)(receiver: Object, position: Object): Object {
    try {
      GenerateStringAt(receiver, position, 'String.prototype.charCodeAt')
          otherwise IfInBounds, IfOutOfBounds;
    }
    label IfInBounds(string: String, index: intptr, length: intptr) {
      const code: int32 = StringCharCodeAt(string, index);
      return Convert<Smi>(code);
    }
    label IfOutOfBounds {
      return kNaN;
    }
  }

  // ES6 #sec-string.prototype.codepointat
  transitioning javascript builtin StringPrototypeCodePointAt(
      implicit context: Context)(receiver: Object, position: Object): Object {
    try {
      GenerateStringAt(receiver, position, 'String.prototype.codePointAt')
          otherwise IfInBounds, IfOutOfBounds;
    }
    label IfInBounds(string: String, index: intptr, length: intptr) {
      // This is always a call to a builtin from Javascript, so we need to
      // produce UTF32.
      const code: int32 = LoadSurrogatePairAt(string, length, index, UTF32);
      return Convert<Smi>(code);
    }
    label IfOutOfBounds {
      return Undefined;
    }
  }

  // ES6 String.prototype.concat(...args)
  // ES6 #sec-string.prototype.concat
  transitioning javascript builtin StringPrototypeConcat(
      implicit context: Context)(receiver: Object, ...arguments): Object {
    // Check that {receiver} is coercible to Object and convert it to a String.
    let string: String = ToThisString(receiver, 'String.prototype.concat');

    // Concatenate all the arguments passed to this builtin.
    const length: intptr = Convert<intptr>(arguments.length);
    for (let i: intptr = 0; i < length; i++) {
      const temp: String = ToString_Inline(context, arguments[i]);
      string = string + temp;
    }
    return string;
  }
}
