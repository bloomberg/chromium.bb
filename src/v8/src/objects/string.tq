// Copyright 2019 the V8 project authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

@abstract
@generateCppClass
@reserveBitsInInstanceType(6)
extern class String extends Name {
  length: int32;
}

@generateCppClass
extern class ConsString extends String {
  first: String;
  second: String;
}

@abstract
extern class ExternalString extends String {
  resource: RawPtr;
  resource_data: RawPtr;
}

extern class ExternalOneByteString extends ExternalString {}
extern class ExternalTwoByteString extends ExternalString {}

@generateCppClass
extern class InternalizedString extends String {
}

// TODO(v8:8983): Add declaration for variable-sized region.
@abstract
@generateCppClass
extern class SeqString extends String {
}
@generateCppClass
extern class SeqOneByteString extends SeqString {
  chars[length]: char8;
}
@generateCppClass
extern class SeqTwoByteString extends SeqString {
  chars[length]: char16;
}

@generateCppClass
extern class SlicedString extends String {
  parent: String;
  offset: Smi;
}

@generateCppClass
extern class ThinString extends String {
  actual: String;
}

// A direct string can be accessed directly through CSA without going into the
// C++ runtime. See also: ToDirectStringAssembler.
type DirectString extends String;
