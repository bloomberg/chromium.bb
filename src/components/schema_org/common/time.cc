// Copyright 2020 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "components/schema_org/common/time.h"

#include <sstream>

namespace schema_org {

base::Optional<base::TimeDelta> ParseISO8601Duration(const std::string& str) {
  if (str.empty() || str[0] != 'P')
    return base::nullopt;

  base::TimeDelta duration;

  std::string time = "";
  int time_index = str.find("T");
  if (time_index == -1)
    return base::nullopt;

  time = str.substr(time_index + 1);
  std::stringstream t(time);
  char unit;
  int amount;

  while (t >> amount) {
    t >> unit;
    switch (unit) {
      case 'H':
        duration = duration + base::TimeDelta::FromHours(amount);
        break;
      case 'M':
        duration = duration + base::TimeDelta::FromMinutes(amount);
        break;
      case 'S':
        duration = duration + base::TimeDelta::FromSeconds(amount);
        break;
      default:
        return base::nullopt;
    }
  }

  return duration;
}

}  // namespace schema_org
