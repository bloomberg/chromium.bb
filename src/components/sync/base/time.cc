// Copyright (c) 2012 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "components/sync/base/time.h"

#include "base/i18n/time_formatting.h"
#include "base/strings/utf_string_conversions.h"

namespace syncer {

int64_t TimeToProtoTime(const base::Time& t) {
  return (t - base::Time::UnixEpoch()).InMilliseconds();
}

base::Time ProtoTimeToTime(int64_t proto_t) {
  return base::Time::UnixEpoch() + base::TimeDelta::FromMilliseconds(proto_t);
}

std::string GetTimeDebugString(const base::Time& t) {
  return base::UTF16ToUTF8(base::TimeFormatFriendlyDateAndTime(t));
}

}  // namespace syncer
