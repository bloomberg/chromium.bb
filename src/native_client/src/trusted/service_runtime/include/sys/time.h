/*
 * Copyright (c) 2012 The Native Client Authors. All rights reserved.
 * Use of this source code is governed by a BSD-style license that can be
 * found in the LICENSE file.
 */

/*
 * NaCl Service Runtime API.  Time types.
 */

#ifndef NATIVE_CLIENT_SRC_TRUSTED_SERVICE_RUNTIME_INCLUDE_SYS_TIME_H_
#define NATIVE_CLIENT_SRC_TRUSTED_SERVICE_RUNTIME_INCLUDE_SYS_TIME_H_

#if !defined(NACL_IN_TOOLCHAIN_HEADERS)
#include "native_client/src/include/build_config.h"
#endif
#include "native_client/src/trusted/service_runtime/include/machine/_types.h"

#ifdef __cplusplus
extern "C" {
#endif

#if defined(NACL_IN_TOOLCHAIN_HEADERS)
#ifndef nacl_abi___susecond_t_defined
#define nacl_abi___susecond_t_defined
typedef long int  nacl_abi_suseconds_t;
#endif
#else
typedef int32_t   nacl_abi_suseconds_t;
#endif

#ifndef nacl_abi___clock_t_defined
#define nacl_abi___clock_t_defined
typedef long int  nacl_abi_clock_t;  /* to be deprecated */
#endif

/*
 * Without this 8-byte alignment, it is possible that the nacl_abi_timeval
 * struct is 12 bytes. However, the equivalent untrusted version of timeval
 * is aligned to 16 bytes. For the utimes function, where an array of
 * "struct timeval" is turned into an array of "struct nacl_abi_timeval",
 * this misalignment can cause an untrusted 32 byte array to be interpreted
 * as a 24 byte trusted array, which is incorrect. This alignment causes
 * the trusted array to be correctly considered 32 bytes in length.
 */
#if defined(NACL_WINDOWS) && NACL_WINDOWS != 0
__declspec(align(8))
#endif
struct nacl_abi_timeval {
  nacl_abi_time_t      nacl_abi_tv_sec;
  nacl_abi_suseconds_t nacl_abi_tv_usec;
#if defined(NACL_WINDOWS) && NACL_WINDOWS != 0
};
#else
} __attribute__((aligned(8)));
#endif

/* obsolete.  should not be used */
struct nacl_abi_timezone {
  int tz_minuteswest;
  int tz_dsttime;
};

/*
 * In some places (e.g., the linux man page) the second parameter is defined
 * as a struct timezone *.  The header file says this struct type should
 * never be used, and defines it by default as void *.  The Mac man page says
 * it is void *.
 */
extern int nacl_abi_gettimeofday(struct nacl_abi_timeval *tv, void *tz);

/*
 * POSIX defined clock id values for clock_getres andn clock_gettime.
 */

#define NACL_ABI_CLOCK_REALTIME           (0)
#define NACL_ABI_CLOCK_MONOTONIC          (1)
#define NACL_ABI_CLOCK_PROCESS_CPUTIME_ID (2)
#define NACL_ABI_CLOCK_THREAD_CPUTIME_ID  (3)

#ifdef __cplusplus
}
#endif

#endif
