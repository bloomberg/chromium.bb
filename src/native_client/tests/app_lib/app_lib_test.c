/*
 * Copyright 2008 The Native Client Authors. All rights reserved.
 * Use of this source code is governed by a BSD-style license that can be
 * found in the LICENSE file.
 */


#include <stdarg.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#include "native_client/src/include/build_config.h"

/*
 * TODO(sehr): enable thread local storage.
 */
#define THREAD

char **environ;

char    bss_buf[1024];  /* just to see ELF info */
int     bss_int;

THREAD int  tls_bss_int;
THREAD char tls_bss_buf[8192];

THREAD int  tls_data_int = 10;
THREAD char tls_data_buf[128] = "Hello world\n";


void  test_tls(void)
{
  printf("%s\n", tls_data_buf);
}


void  test_fmt(char *fmt, int val)
{
  fputs(fmt, stdout); putc(':', stdout);
  printf(fmt, val);
  putc('\n', stdout);
}

void  test_fp_fmt(char  *fmt, double val)
{
  fputs(fmt, stdout); putc(':', stdout);
  printf(fmt, val);
  putc('\n', stdout);
}

void  test_fp_val(double  val)
{
  printf("\n123456789012345678901234567890"
         "123456789012345678901234567890\n");
  test_fp_fmt("%e", val);
  test_fp_fmt("%30.15e.", val);
  test_fp_fmt("%-30.15e.", val);
  test_fp_fmt("%f.", val);
  test_fp_fmt("%30.15f.", val);
  test_fp_fmt("%-30.15f.", val);
}

void  test_fp(void)
{
  double d;

  printf("one half is %e\n", 0.5);
  printf("one half is %f\n", 0.5);
  printf("one third is %e\n", 1.0/3.0);
  printf("one third is %f\n", 1.0/3.0);

  test_fp_val(0.0);
  test_fp_fmt("%.500e", 1.0);
  d = 1.234567e+10;
  test_fp_val(d);
  test_fp_val(-d);

  d = 1.234567e308;
  test_fp_val(d);
  test_fp_val(-d);

  d = 1.234567e-307;
  test_fp_val(d);
  test_fp_val(-d);

  d = 1.234567e-310;
  test_fp_val(d);
  test_fp_val(-d);
#if 0
  /* TODO: the code below doesn't compile on Windows
     NOTE(robertm): disabled this code for all platforms
     to be able to do with one golden file */
  d = 1.234567e+310;  /* inf */
  test_fp_val(d);
  test_fp_val(-d);
  d = 1.234567e+310;  /* inf */
  test_fp_val(d);
  test_fp_val(-d);

  d = 0.0/0.0;  /* nan */
  test_fp_val(d);
#endif  /*NACL_WINDOWS*/
}

void  test_malloc(void)
{
  char    *p;
  printf("test_malloc: entered\n");
  p = malloc(10);
  fprintf(stderr, "# got 0x%08x\n", (unsigned int) p);
  printf("test_malloc: leaving\n");
}

void test_time(void)
{
  time_t tval;
  printf("test_time: entered\n");
  tval = time(NULL);
  fprintf(stderr, "# current time is %d \n", (int) tval);
  printf("test_time: leaving\n");
}

void  test_fopen(const char* fn)
{
  FILE  *fp;
  int   c;
  char  line[1024];
  char  *rv;

  printf("test_fopen: entered\n");
  fp = fopen(fn, "rb");
  if (NULL == fp) {
    printf("could not open file %s!\n", fn);
    exit(-1);
  } else {
    printf("reading file\n");
    while (EOF != (c = fgetc(fp))) {
      if ((int)c != 13)
        /*
          TODO(gregoryd): make sure the input files are identical
          on all platforms, so we won't need the if.
        */
        putc(c, stdout);
    }
    fclose(fp);
  }

  while (0 != (rv = (fputs("input: ", stdout),
                     fflush(stdout),
                     fgets(line, sizeof line, stdin)))) {
    printf("Got line: <<%s>>, %d\n", rv, strlen(rv));
  }
  printf("test_fopen: done\n");
}


int NaClMain(const char* fn)
{
  char const  *hello = "Hello world\n";
  char const  *bye = "Goodbye cruel world";
  int         r;

  r = write(1, hello, strlen(hello));

  printf("write returned %d.\n", r);
  test_fmt("<%04d>", r);
  test_fmt("<%-d>", r);
  test_fmt("<%-04d>", r);
  r = -r;
  test_fmt("<%d>", r);
  test_fmt("<%04d>", r);
  test_fmt("<%-d>", r);
  test_fmt("<%-04d>", r);

  printf("The message was \"%s\".\n", hello);

  printf("Bye message: \"%25s\".\n", bye);
  printf("Bye message: \"%-25s\".\n", bye);

  test_fp();
  test_malloc();
  test_time();
  test_fopen(fn);

  fflush(0);
  fprintf(stderr, "# Hello is at address 0x%08lx\n", (long) hello);
  fprintf(stderr, "# Hello is at address %p\n", (void *) hello);
  {
    short s = -1;
    unsigned short us = (unsigned short) -1;

    printf("s = %hd\n", s);
    printf("us = %hu\n", us);
  }

  printf("Goodbye cruel world.  Really.\n");
  fprintf(stderr, "My pid is %d\n", getpid());
  fflush(0);
  return 0;
}

int SimpleMain(void)
{
  static char const  hello[] = "Hello from SimpleMain\n";

  write(1, hello, -1 + sizeof(hello));
  return 0;
}

int main(int argc, char *argv[], char **ep)
{
  if (argc < 2) {
    printf("you must specify a file to read\n");
    exit(-1);
  }
#if 1
  return NaClMain(argv[1]);
#else
  return SimpleMain();
#endif
}

void test_start(char *arg, ...)
{
  int   argc;
  char  **argv;
  char  **envp;

  argv = &arg;
  argc = *(int *)(argv-1);
  envp = argv + argc + 1;

  environ = envp;

  exit(main(argc, argv, envp));
}
