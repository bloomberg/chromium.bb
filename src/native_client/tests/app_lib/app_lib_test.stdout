Hello world
write returned 12.
<%04d>:<0012>
<%-d>:<12>
<%-04d>:<12  >
<%d>:<-12>
<%04d>:<-012>
<%-d>:<-12>
<%-04d>:<-12 >
The message was "Hello world
".
Bye message: "      Goodbye cruel world".
Bye message: "Goodbye cruel world      ".
one half is 5.000000e-01
one half is 0.500000
one third is 3.333333e-01
one third is 0.333333

123456789012345678901234567890123456789012345678901234567890
%e:0.000000e+00
%30.15e.:         0.000000000000000e+00.
%-30.15e.:0.000000000000000e+00         .
%f.:0.000000.
%30.15f.:             0.000000000000000.
%-30.15f.:0.000000000000000             .
%.500e:1.00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e+00

123456789012345678901234567890123456789012345678901234567890
%e:1.234567e+10
%30.15e.:         1.234567000000000e+10.
%-30.15e.:1.234567000000000e+10         .
%f.:12345670000.000000.
%30.15f.:   12345670000.000000000000000.
%-30.15f.:12345670000.000000000000000   .

123456789012345678901234567890123456789012345678901234567890
%e:-1.234567e+10
%30.15e.:        -1.234567000000000e+10.
%-30.15e.:-1.234567000000000e+10        .
%f.:-12345670000.000000.
%30.15f.:  -12345670000.000000000000000.
%-30.15f.:-12345670000.000000000000000  .

123456789012345678901234567890123456789012345678901234567890
%e:1.234567e+308
%30.15e.:        1.234567000000000e+308.
%-30.15e.:1.234567000000000e+308        .
%f.:123456700000000005208528432756090151921261274161788916303676059705979462063742969732144909533540519431855799682461437884122008361354180847630642882297863747287304471927156665386555144738399912230154762806869335545190224118447554993673509095546955604425539464758125498720242123200518548236068402284934476070912.000000.
%30.15f.:123456700000000005208528432756090151921261274161788916303676059705979462063742969732144909533540519431855799682461437884122008361354180847630642882297863747287304471927156665386555144738399912230154762806869335545190224118447554993673509095546955604425539464758125498720242123200518548236068402284934476070912.000000000000000.
%-30.15f.:123456700000000005208528432756090151921261274161788916303676059705979462063742969732144909533540519431855799682461437884122008361354180847630642882297863747287304471927156665386555144738399912230154762806869335545190224118447554993673509095546955604425539464758125498720242123200518548236068402284934476070912.000000000000000.

123456789012345678901234567890123456789012345678901234567890
%e:-1.234567e+308
%30.15e.:       -1.234567000000000e+308.
%-30.15e.:-1.234567000000000e+308       .
%f.:-123456700000000005208528432756090151921261274161788916303676059705979462063742969732144909533540519431855799682461437884122008361354180847630642882297863747287304471927156665386555144738399912230154762806869335545190224118447554993673509095546955604425539464758125498720242123200518548236068402284934476070912.000000.
%30.15f.:-123456700000000005208528432756090151921261274161788916303676059705979462063742969732144909533540519431855799682461437884122008361354180847630642882297863747287304471927156665386555144738399912230154762806869335545190224118447554993673509095546955604425539464758125498720242123200518548236068402284934476070912.000000000000000.
%-30.15f.:-123456700000000005208528432756090151921261274161788916303676059705979462063742969732144909533540519431855799682461437884122008361354180847630642882297863747287304471927156665386555144738399912230154762806869335545190224118447554993673509095546955604425539464758125498720242123200518548236068402284934476070912.000000000000000.

123456789012345678901234567890123456789012345678901234567890
%e:1.234567e-307
%30.15e.:        1.234567000000000e-307.
%-30.15e.:1.234567000000000e-307        .
%f.:0.000000.
%30.15f.:             0.000000000000000.
%-30.15f.:0.000000000000000             .

123456789012345678901234567890123456789012345678901234567890
%e:-1.234567e-307
%30.15e.:       -1.234567000000000e-307.
%-30.15e.:-1.234567000000000e-307       .
%f.:-0.000000.
%30.15f.:            -0.000000000000000.
%-30.15f.:-0.000000000000000            .

123456789012345678901234567890123456789012345678901234567890
%e:1.234567e-310
%30.15e.:        1.234567000000017e-310.
%-30.15e.:1.234567000000017e-310        .
%f.:0.000000.
%30.15f.:             0.000000000000000.
%-30.15f.:0.000000000000000             .

123456789012345678901234567890123456789012345678901234567890
%e:-1.234567e-310
%30.15e.:       -1.234567000000017e-310.
%-30.15e.:-1.234567000000017e-310       .
%f.:-0.000000.
%30.15f.:            -0.000000000000000.
%-30.15f.:-0.000000000000000            .
test_malloc: entered
test_malloc: leaving
test_time: entered
test_time: leaving
test_fopen: entered
reading file
test
stdin
test
input: Got line: <<test
>>, 5
input: Got line: <<stdin
>>, 6
input: Got line: <<test
>>, 5
input: test_fopen: done
s = -1
us = 65535
Goodbye cruel world.  Really.
